cirq-core/cirq/contrib/bayesian_network/bayesian_network_gate.py

```diff
@@ -0,0 +1,198 @@
+            ('q1', 0.456)   # Indicates that p(q1 = 1) = 0.456
+            ('q2', None)    # Indicates that q2 is a dependent variable
+        ]
+        arc_probs = [
+            ('q2', ('q0', 'q1'), [0.1, 0.2, 0.3, 0.4])
+                # Indicates that p(q2 = 1 | q0 = 0 and q1 = 0) = 0.1
+                # Indicates that p(q2 = 1 | q0 = 0 and q1 = 1) = 0.2
+                # Indicates that p(q2 = 1 | q0 = 1 and q1 = 0) = 0.3
+                # Indicates that p(q2 = 1 | q0 = 1 and q1 = 1) = 0.4
+        ]
+
+        By convention, all the probabilties are for the variable being equal to 1 and the
+        probability of being equal to zero can be inferred. In the example above, we thus have:
+        p(q2 = 0 | q0 = 1 and q1 = 0) = 1.0 - p(q2 = 1 | q0 = 1 and q1 = 0) = 0.7
+
+        Note that there is NO checking that the chain of probability creates a directed acyclic
+        graph. In particular, the order of the elements in arc_probs matters. Also, if you want to
+        specify the dependent probabilities outside of this gate, you can mark all the variables as
+        dependent in init_probs.
+
+        init_prob: A list of tuples, each representing a single variable. The first element of the
+            tuples is a string representing the name of the variable. The second element of the
+            tuples is either None for dependent variables, or a float representing a probability.
+
+        arc_probs: A list of tuples, each representing a dependence. The first element of the tuples
+            is a string representing the name of the variable. The second element of the tuples is
+            itself a tuple of n strings, representing the dependence. The third element of the
+            tuples is a list of 2**n floats, each representing the probabilities.
+        """
+        for _, init_prob in init_probs:
+            if init_prob is None:
+                continue
+            if init_prob < 0.0 or init_prob > 1.0:
+                raise ValueError('Initial prob should be between 0 and 1.')
+        self._init_probs = init_probs
+        for _, params, cond_probs in arc_probs:
+            if not isinstance(params, tuple):
+                raise ValueError('Conditional prob params must be a tuple.')
+            if len(cond_probs) != 2 ** len(params):
+                raise ValueError('Incorrect number of conditional probs.')
+            for cond_prob in cond_probs:
+                if cond_prob < 0.0 or cond_prob > 1.0:
+                    raise ValueError('Conditional prob should be between 0 and 1.')
+        self._arc_probs = arc_probs
+
+    def _decompose_(self, qubits: Sequence['cirq.Qid']) -> 'cirq.OP_TREE':
+        parameter_names = [init_prob[0] for init_prob in self._init_probs]
+        qubit_map = dict(zip(parameter_names, qubits))
+
+        for param, init_prob in self._init_probs:
+            yield self._generate_got_set_for_init_prob(qubit_map[param], init_prob)
```

Comment:
Are we assuming all nodes are root nodes and none follow along with eqn 14 from the paper ?

Reply:
Correct. We just specify the dependent nodes with a `None` probability. Also the user messes up and specify a node is independent (i.e. with an initial probability) and then assigns it as a dependent variable, the code will run fine (but return absurd results).

The reason I decided to do it this way, was that they are other ways to mess things up. We would have to check the graph is directed acyclic, but that seems like pulling a lot of graph code in. I can do it if you want, of course. For now, I just put a warning in the docstring.

Leaving unresolved. Let me know?